src/libraries/System.Private.CoreLib/src/System/Collections/Generic/NonRandomizedStringEqualityComparer.cs

```diff
@@ -108,12 +108,12 @@ internal override RandomizedStringEqualityComparer GetRandomizedEqualityComparer
             }
         }
 
-        public static IEqualityComparer<string>? GetStringComparer(object? comparer)
+        public static IEqualityComparer<string>? GetStringComparer(object comparer)
         {
             // Special-case EqualityComparer<string>.Default, StringComparer.Ordinal, and StringComparer.OrdinalIgnoreCase.
             // We use a non-randomized comparer for improved perf, falling back to a randomized comparer if the
             // hash buckets become unbalanced.
-            if (comparer is null)
+            if (ReferenceEquals(comparer, EqualityComparer<string>.Default))
```

Comment:
I believe that there are cases in binary serialization where we end with multiple instances of the comparer. It is why EqualityComparers override Equals(object). I am not sure whether we can hit those cases here, but it would be safer to call `object.Equals(object)` here.

Reply:
Somewhat related: This comment https://github.com/dotnet/runtime/blob/4a04e5d9b23e204ceac6e1f4c1621a4738fa80a2/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/EqualityComparer.cs#L133 can be addressed now - the type is sealed now.